### DISCUSSION TITLE
How to add meta tags which I first need to fetch

### DISCUSSION THREAD

**Discussion by Borisboky** [gatsbyjs/gatsby#35449] [Help]
I'm trying to og-tags with Gatsby and Helmet. But the problem is that those tags first need to be fetched.

It's a vehicle detail page and I need to show vehicle make and model in those tags, but the vehicle needs first to be fecthed. My code is as follows:

```
const Page = (props) => {
    const [detailsMeta, setDetailsMeta] = useState(undefined);


    const resolveVehicleDetailMeta = async () => {
          if(props?.reference){
              const fetch = require('isomorphic-fetch');
    
              const resolveVehicleImageUrl = (fetchedImage) => {
                const parsed = JSON.parse(fetchedImage);
                return parsed?.uri
              }
    
              const VEHICLE_QUERY = `
                query VehicleQuery($reference: String!) {
                  vehicle (reference: $reference) {
                    reference
                    make
                    model
                    image
                  }
                }`;
    
              await fetch(`/graphql`, {
                method: 'POST',
                headers: { 'Content-Type': 'application/json' },
                body: JSON.stringify({
                  query: VEHICLE_QUERY,
                  variables: {
                    reference: 'some reference'
                  }
                })
              })
            .then((resp) => resp.json())
            .then((result) => {
              const vehicle = result?.data?.vehicle;
              if(vehicle){
                setDetailsMeta({
                  reference: vehicle.reference,
                  make: vehicle.make,
                  model: vehicle.model,
                  image: resolveVehicleImageUrl(vehicle.image)
                })
    
              }
            })
            .catch((err) => {
              console.log('err', err)
            });
        }
    }

    const renderMetaTitle = () => {
      const ogTitle = `Tweedehands ${detailsMeta?.make} ${detailsMeta?.model} ${detailsMeta?.reference}`

      return ogTitle;
    }

    return (
        <>
            <Helmet>
                {detailsMeta && <meta property='og:title' content={renderMetaTitle()} />}
                ...
            </Helmet>

            The rest...
        </>
    )



}
```

Thus, I first fetch data from the server and store it in detailsMeta and then I show it inside Helmet. When I test it on localhost I see those tags and it works fine, but when I test it in Facebook debugger they are not shown.

Can the data which first needs to be fetched be added at all?

Thanks.

**Comment by ratkorle** [comment]
You can create the page from gatsby-node.js do the data fetching there and pass the data to your page and you will have all required SEO data for sure.

https://www.gatsbyjs.com/docs/reference/config-files/gatsby-node/#createPages

**Reply by Borisboky** [reply]
Thanks for your reply.

The problem is that I have only one detail page and not a page for every vehicle. And in my case it's not possible to do so. When I render Page I first need to fetch the vehicle details.

In other words, I cant do the fetching in CreatePages because I do not have a vehicle reference there. I can fetch all the vehicles and create those meta tags for every vehicle, but on Page I need to know which one I need to show and that can I do only if I have `props?.reference` which I have after load. And because of that those tags are still not there on first load.
